cat/riscv.cat

```diff
@@ -0,0 +1,87 @@
+Partial
+
+(***************)
+(* Definitions *)
+(***************)
+
+(*************)
+(* Utilities *)
+(*************)
+
+let fence.r.r = [R];fencerel(Fence.r.r);[R]
+let fence.r.w = [R];fencerel(Fence.r.w);[W]
+let fence.r.rw = [R];fencerel(Fence.r.rw);[M]
+let fence.w.r = [W];fencerel(Fence.w.r);[R]
+let fence.w.w = [W];fencerel(Fence.w.w);[W]
+let fence.w.rw = [W];fencerel(Fence.w.rw);[M]
+let fence.rw.r = [M];fencerel(Fence.rw.r);[R]
+let fence.rw.w = [M];fencerel(Fence.rw.w);[W]
+let fence.rw.rw = [M];fencerel(Fence.rw.rw);[M]
+let fence.tso = ([W];fencerel(Fence.tso);[W]) | ([R];fencerel(Fence.tso);[M])
+
+let fence = 
+  fence.r.r | fence.r.w | fence.r.rw |
+  fence.w.r | fence.w.w | fence.w.rw |
+  fence.rw.r | fence.rw.w | fence.rw.rw |
+  fence.tso
+
+
+let po-loc-no-w = po-loc \ (po-loc?;[W];po-loc)
+let rsw = rf^-1;rf
+let AcqRel = AcqRel|Sc (* Compat *)
+let AQ = (Acq|AcqRel)
+let RL = (Rel|AcqRel)
+//let AMO = try AMO with (R & W) (* Compat *)
+let RCsc = (Acq|Rel|AcqRel) & (AMO|X)
+(*************)
+(* ppo rules *)
+(*************)
+
+(* Overlapping-Address Orderings *)
+let r1 = [M];po-loc;[W]
+let r2 = ([R];po-loc-no-w;[R]) \ rsw
+let r3 = [AMO|X];rfi;[R]
+(* Explicit Synchronization *)
+let r4 = fence
+let r5 = [AQ];po;[M]
+let r6 = [M];po;[RL]
+let r7 = [RCsc];po;[RCsc]
+let r8 = rmw
```

Comment:
As we once discussed, I'll create a new `RelCopy` derived relation which depends on the original one. But this is more related to CAAT than with RISCV so I'll do this in a separate PR (and I might need your help to implement the corresponding graph + whatever else is needed in CAAT).

Reply:
I'm honestly not too sure that it is related to `CAAT` in particular. A relation can only have one name, so what is the name of `RelRMW` in that case? Is it still `rmw` or will it get overwritten by `r8`? The way we store relations simply does not really allow to have aliases for the same relation IIRC.

Reply:
Each `let` creates a new relations. Thus we will have two relations (one called `rmw` and one called `r8`) that happen to behave identically.

Reply:
I don't think so:
```    
@Override
    public Object visitLetDefinition(CatParser.LetDefinitionContext ctx) {
        Relation r = ctx.e.accept(relationVisitor); // Parses the rhs, in this case "rmw" and hence returns `RelRMW`
        if(r != null){
            r.setName(ctx.n.getText()); // Changes the name of `rmw` to `r8`
            relationRepository.updateRelation(r);
        } else {
            // ... Filter stuff
        }
        return null;
    }
```
If it would create a new relation, of what type would it even be? It cannot be `UnaryRelation` nor `BinaryRelation`. It could at best be a fresh copy of `RelRMW`

Reply:
The code might not be creating new relations right now, but I think it is what it should do (modulo (-) below).
Why couldn't it return a new (extension of) `UnaryRelation `? Something pretty close to `RelInverse` but without inverting the tuples.
(-) One would have to be cautious to call this "create copy" when having `let r1 = rmw`, but not `let r1 = r0 | rmw` because here we effectively want `rmw` and not a copy.

Reply:
Sure, we can have a new copy relation. All I said is that right now, a let definition `let r8 = rmw` is inherently broken and it is not tied to just `CAAT`.
What I'm thinking might be the best way, is to actually let relations have multiple names (aliases). Having an extra relation in the `Wmm` that is effectively just identical to another relation will likely cause various problems down the line. In particular, the copy relation will be an `UnaryRelation` with a dependency that is different to the original relation. This is bound to cause issues where the original relation and the copy are treated differently, or it forces us to add special case reasoning everywhere.
That being said, we can add aliases in a different PR.

Reply:
Using aliases seems to be the easiest/cleanest.

I'll change `r8` to `rmw` and implement aliasing in a different PR.

Reply:
sounds fair to me

Reply:
Since this is the only unresolved issue and the CI passed, can we merge?